compute-hypothetical-feature.mjs

```diff
@@ -0,0 +1,143 @@
+
+const caniuse = JSON.parse(readFileSync('caniuse/fulldata-json/data-2.0.json', 'utf8'));
+const historicalBrowserData = JSON.parse(readFileSync('historical-browser-data.json', 'utf8'));
+
+// Sort the data in |historicalBrowserData| so that we can search it by date later.
+for (const snapshots of Object.values(historicalBrowserData)) {
+    snapshots.sort((a, b) => a.timestamp - b.timestamp);
+}
+
+const browsers = ['chrome', 'edge', 'firefox', 'safari'];
+
+const endDate = new Date();
+
+// Pre-process |a| and |b| to normalize some caniuse versions.
+function caniuseCompare(a, b, cmp) {
+    function normalize(version) {
+        if (version == 'TP') {
+            // Treat Safari TP as a future release. It doesn't matter because
+            // it has no usage in the caniuse data.
+            return String(Number.MAX_SAFE_INTEGER);
+        }
+        if (version == '15.2-15.3') {
+            return '15.2';
+        }
+        return version;
+    }
+    return compare(normalize(a), normalize(b), cmp);
+}
+
+// timestamp is number of seconds since epoch
+function caniuseDate(timestamp) {
+  return new Date(1000 * timestamp);
+}
+
+// from (inclusive) and to (exclusive) are Date objects.
+// TODO: this isn't quite right, date 
+function* iterateMonths(from, to) {
+    const iter = new Date(from);
+    while (iter < to) {
+        yield new Date(iter);
+        // Add one calendar month.
+        const nextMonth = (iter.getUTCMonth() + 1) % 12;
+        if (nextMonth == 0) {
+            iter.setUTCFullYear(iter.getUTCFullYear() + 1);
+        }
+        iter.setUTCMonth(nextMonth);
+    }
+}
+
+function monthString(date) {
+    return date.toISOString().substr(0, 10);
+}
+
+// Find the usage of a browser at a given date. The exact date is probably not
+// found in |historicalBrowserData|, so use the latest entry at or before
+// |date|. If there is none, return the very latest entry. The caller needs to
+// check that the data is fresh enough before using it.
+function getHistoricalUsage(browser, date) {
+    const snapshots = historicalBrowserData[browser];
+    // TODO: binary search would be O(log(n)) instead of O(n).
+    const index = snapshots.findLastIndex((s) => caniuseDate(s.timestamp) <= date);
+    // Index could be -1, and we return the latest snapshot in this case.
+    return snapshots.at(index);
+}
+
+const months = Array.from(iterateMonths(Date.parse('2020-01-01T00:00Z'), endDate));
+
+function getVersionAfter(browser, date) {
+    for (const release of caniuse.agents[browser].version_list) {
+        if (!release.release_date) {
+            continue;
+        }
+        const releaseDate = caniuseDate(release.release_date);
+        if (releaseDate >= date) {
+            return release.version;
+        }
+    }
+    return null;
+}
+
+for (const startIndex of months.keys()) {
+    const launchDate = months[startIndex];
+    // Assume that a hypothetical feature is enabled in all browser engines at
```

Comment:
@dfabulich if you want this script, I can write something in the README instead.